// node-opcua-address-space/dist/src/state_machine/ua_two_state_variable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._addTwoStateVariable = exports.UATwoStateVariable = exports.promoteToTwoStateVariable = exports._install_TwoStateVariable_machinery = void 0;
/**
 * @module node-opcua-address-space
 */
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_constants_1 = require("node-opcua-constants");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const node_opcua_nodeid_2 = require("node-opcua-nodeid");
const node_opcua_status_code_1 = require("node-opcua-status-code");
const node_opcua_variant_1 = require("node-opcua-variant");
const node_opcua_variant_2 = require("node-opcua-variant");
const ua_variable_1 = require("../ua_variable");
const register_node_promoter_1 = require("../../source/loader/register_node_promoter");
const hasTrueSubState_ReferenceTypeNodeId = node_opcua_nodeid_1.resolveNodeId("HasTrueSubState");
const hasFalseSubState_ReferenceTypeNodeId = node_opcua_nodeid_1.resolveNodeId("HasFalseSubState");
// Release 1.03 12 OPC Unified Architecture, Part 9
// Two-state state machines
// Most states defined in this standard are simple – i.e. they are either TRUE or FALSE. The
// TwoStateVariableType is introduced specifically for this use case. More complex states are
// modelled by using a StateMachineType defined in Part 5.
// The TwoStateVariableType is derived from the StateVariableType.
//
// Attribute        Value
// BrowseName       TwoStateVariableType
// DataType         LocalizedText
// ValueRank        -1 (-1 = Scalar)
// IsAbstract       False
//
// Subtype of the StateVariableType defined in Part 5.
// Note that a Reference to this subtype is not shown in the definition of the StateVariableType
//
// References      NodeClass BrowseName              DataType      TypeDefinition Modelling Rule
// HasProperty     Variable  Id                      Boolean       PropertyType   Mandatory
// HasProperty     Variable  TransitionTime          UtcTime       PropertyType   Optional
// HasProperty     Variable  EffectiveTransitionTime UtcTime       PropertyType   Optional
// HasProperty     Variable  TrueState               LocalizedText PropertyType   Optional
// HasProperty     Variable  FalseState              LocalizedText PropertyType   Optional
// HasTrueSubState StateMachine or
//                 TwoStateVariableType
//                                                  <StateIdentifier> Defined in Clause 5.4.2 Optional
// HasFalseSubState StateMachine or
//                  TwoStateVariableType
//                                                  <StateIdentifier> Defined in Clause 5.4.3 Optional
function _updateTransitionTime(node) {
    // TransitionTime specifies the time when the current state was entered.
    if (node.transitionTime) {
        node.transitionTime.setValueFromSource({ dataType: node_opcua_variant_2.DataType.DateTime, value: new Date() });
    }
}
function _updateEffectiveTransitionTime(node) {
    if (node.effectiveTransitionTime) {
        // xx console.log("xxxx _updateEffectiveTransitionTime
        // because subStateNode ",subStateNode.browseName.toString());
        node.effectiveTransitionTime.setValueFromSource({
            dataType: node_opcua_variant_2.DataType.DateTime,
            value: new Date()
        });
    }
}
function _getEffectiveDisplayName(node) {
    const humanReadableString = _getHumanReadableString(node);
    if (humanReadableString.statusCode !== node_opcua_status_code_1.StatusCodes.Good) {
        return humanReadableString;
    }
    const boolValue = node.getValue();
    let subStateNodes;
    if (boolValue) {
        subStateNodes = node.findReferencesExAsObject("HasTrueSubState", node_opcua_data_model_1.BrowseDirection.Forward);
    }
    else {
        subStateNodes = node.findReferencesExAsObject("HasFalseSubState", node_opcua_data_model_1.BrowseDirection.Forward);
    }
    const states = subStateNodes.forEach((n) => {
        // todo happen
    });
    return humanReadableString;
}
function _getHumanReadableString(node) {
    let dataValue = node.id.readValue();
    if (dataValue.statusCode !== node_opcua_status_code_1.StatusCodes.Good) {
        const _c = dataValue.clone();
        _c.value = new node_opcua_variant_1.Variant({
            dataType: node_opcua_variant_2.DataType.LocalizedText,
            value: node_opcua_data_model_1.coerceLocalizedText("")
        });
        return _c;
    }
    node_opcua_assert_1.assert(dataValue.value.dataType === node_opcua_variant_2.DataType.Boolean);
    const boolValue = dataValue.value.value;
    // The Value Attribute of a TwoStateVariable contains the current state as a human readable name.
    // The EnabledState for example, might contain the name “Enabled” when TRUE and “Disabled” when FALSE.
    const dataValue2 = dataValue.clone();
    dataValue2.value = new node_opcua_variant_1.Variant({
        dataType: node_opcua_variant_2.DataType.LocalizedText,
        value: boolValue ? node.getTrueState() : node.getFalseState()
    });
    return dataValue2;
}
function _install_TwoStateVariable_machinery(node, options) {
    node_opcua_assert_1.assert(node.dataTypeObj.browseName.toString() === "LocalizedText");
    node_opcua_assert_1.assert(node.minimumSamplingInterval === 0);
    node_opcua_assert_1.assert(node.typeDefinitionObj.browseName.toString() === "TwoStateVariableType");
    node_opcua_assert_1.assert(node.dataTypeObj.browseName.toString() === "LocalizedText");
    node_opcua_assert_1.assert(node.hasOwnProperty("valueRank") && (node.valueRank === -1 || node.valueRank === 0));
    // promote node into a UATwoStateVariable
    const _node = promoteToTwoStateVariable(node);
    node.initialize(options);
    return node;
}
exports._install_TwoStateVariable_machinery = _install_TwoStateVariable_machinery;
function promoteToTwoStateVariable(node) {
    if (node instanceof UATwoStateVariable) {
        return node;
    }
    // istanbul ignore next
    if (!(node instanceof ua_variable_1.UAVariable)) {
        throw new Error("Trying to promote a invalid object");
    }
    Object.setPrototypeOf(node, UATwoStateVariable.prototype);
    return node;
}
exports.promoteToTwoStateVariable = promoteToTwoStateVariable;
register_node_promoter_1.registerNodePromoter(node_opcua_constants_1.VariableTypeIds.TwoStateVariableType, promoteToTwoStateVariable);
/***
 * @class UATwoStateVariable
 * @constructor
 * @extends UAVariable
 */
class UATwoStateVariable extends ua_variable_1.UAVariable {
    constructor(opts) {
        super(opts);
    }
    get isFalseSubStateOf() {
        return super.isFalseSubStateOf;
    }
    get isTrueSubStateOf() {
        return super.isTrueSubStateOf;
    }
    initialize(options) {
        const node = this;
        if (options.trueState) {
            node_opcua_assert_1.assert(!!options.falseState);
            node_opcua_assert_1.assert(typeof options.trueState === "string");
            node_opcua_assert_1.assert(typeof options.falseState === "string");
            if (node.falseState) {
                node.falseState.setValueFromSource({
                    dataType: node_opcua_variant_2.DataType.LocalizedText,
                    value: node_opcua_data_model_1.coerceLocalizedText(options.falseState)
                });
            }
            else {
                node._trueState = node_opcua_data_model_1.coerceLocalizedText(options.trueState).text;
            }
            if (node.trueState) {
                node.trueState.setValueFromSource({
                    dataType: node_opcua_variant_2.DataType.LocalizedText,
                    value: node_opcua_data_model_1.coerceLocalizedText(options.trueState)
                });
            }
            else {
                node._falseState = node_opcua_data_model_1.coerceLocalizedText(options.falseState).text;
            }
        }
        // handle isTrueSubStateOf
        if (options.isTrueSubStateOf) {
            node.addReference({
                isForward: false,
                nodeId: options.isTrueSubStateOf,
                referenceType: "HasTrueSubState"
            });
        }
        if (options.isFalseSubStateOf) {
            node.addReference({
                isForward: false,
                nodeId: options.isFalseSubStateOf,
                referenceType: "HasFalseSubState"
            });
        }
        if (options.value === undefined) {
            node.id.setValueFromSource({
                dataType: "Boolean",
                value: false
            }, node_opcua_status_code_1.StatusCodes.UncertainInitialValue);
        }
        else if (typeof options.value === "boolean") {
            node.id.setValueFromSource({
                dataType: "Boolean",
                value: options.value
            }, node_opcua_status_code_1.StatusCodes.Good);
        }
        else if (options.value.hasOwnProperty("dataType")) {
            node_opcua_assert_1.assert(options.value.dataType === node_opcua_variant_2.DataType.Boolean);
            node.id.setValueFromSource(options.value, node_opcua_status_code_1.StatusCodes.Good);
        }
        else {
            node.id.bindVariable(options.value);
        }
        this._postInitialize();
    }
    _postInitialize() {
        const node = this;
        if (node.effectiveTransitionTime) {
            // install "value_changed" event handler on SubState that are already defined
            const subStates = [].concat(node.getTrueSubStates(), node.getFalseSubStates());
            for (const subState of subStates) {
                subState.on("value_changed", () => _updateEffectiveTransitionTime(node));
            }
        }
        // it should be possible to define a trueState and falseState LocalizedText even if
        // the trueState or FalseState node is not exposed. Therefore we need to store their value
        // into dedicated variables.
        node.id.on("value_changed", () => {
            node._internal_set_dataValue(_getHumanReadableString(node));
        });
        node._internal_set_dataValue(_getHumanReadableString(node));
        // todo : also set the effectiveDisplayName if present
        // from spec Part 5
        // Release 1.03 OPC Unified Architecture, Part 5
        // EffectiveDisplayName contains a human readable name for the current state of the state
        // machine after taking the state of any SubStateMachines in account. There is no rule specified
        // for which state or sub-state should be used. It is up to the Server and will depend on the
        // semantics of the StateMachineType
        //
        // EffectiveDisplayName will be constructed by adding the EnableState
        // and the State of the addTrue state
        if (node.effectiveDisplayName) {
            node.id.on("value_changed", () => {
                node.effectiveDisplayName._internal_set_dataValue(_getEffectiveDisplayName(node));
            });
            node.effectiveDisplayName._internal_set_dataValue(_getEffectiveDisplayName(node));
        }
    }
    /**
     * @method setValue
     * @param boolValue {Boolean}
     */
    setValue(boolValue) {
        const node = this;
        node_opcua_assert_1.assert(typeof boolValue === "boolean");
        const dataValue = node.id.readValue();
        const oldValue = dataValue.value.value;
        if (dataValue.statusCode === node_opcua_status_code_1.StatusCodes.Good && boolValue === oldValue) {
            return; // nothing to do
        }
        //
        node.id.setValueFromSource(new node_opcua_variant_1.Variant({ dataType: node_opcua_variant_2.DataType.Boolean, value: boolValue }));
        _updateTransitionTime(node);
        _updateEffectiveTransitionTime(node);
    }
    /**
     * @method getValue
     * @return {Boolean}
     */
    getValue() {
        const node = this;
        const dataValue = node.id.readValue();
        node_opcua_assert_1.assert(dataValue.statusCode === node_opcua_status_code_1.StatusCodes.Good);
        node_opcua_assert_1.assert(dataValue.value.dataType === node_opcua_variant_2.DataType.Boolean);
        return dataValue.value.value;
    }
    /**
     * @method getValueAsString
     * @return {string}
     */
    getValueAsString() {
        const node = this;
        const dataValue = node.readValue();
        node_opcua_assert_1.assert(dataValue.statusCode === node_opcua_status_code_1.StatusCodes.Good);
        node_opcua_assert_1.assert(dataValue.value.dataType === node_opcua_variant_2.DataType.LocalizedText);
        return dataValue.value.value.text.toString();
    }
    getTrueState() {
        return this.trueState ? this.trueState.readValue().value.value : node_opcua_data_model_1.coerceLocalizedText(this._trueState || "TRUE");
    }
    getFalseState() {
        return this.falseState ? this.falseState.readValue().value.value : node_opcua_data_model_1.coerceLocalizedText(this._falseState || "FALSE");
    }
    // TODO : shall we care about overloading the remove_backward_reference method ?
    // some TrueSubState and FalseSubState relationship may be added later
    // so we need a mechanism to keep adding the "value_changed" event handle on subStates that
    // will be defined later.
    // install change detection on sub State
    // this is useful to change the effective transitionTime
    // EffectiveTransitionTime specifies the time when the current state or one of its sub states was entered.
    // If, for example, a LevelAlarm is active and – while active – switches several times between High and
    // HighHigh, then the TransitionTime stays at the point in time where the Alarm became active whereas the
    // EffectiveTransitionTime changes with each shift of a sub state.
    _add_backward_reference(reference) {
        const self = this;
        super._add_backward_reference(reference);
        if (reference.isForward &&
            (node_opcua_nodeid_2.sameNodeId(reference.referenceType, hasTrueSubState_ReferenceTypeNodeId) ||
                node_opcua_nodeid_2.sameNodeId(reference.referenceType, hasFalseSubState_ReferenceTypeNodeId))) {
            const addressSpace = self.addressSpace;
            // add event handle
            const subState = addressSpace.findNode(reference.nodeId);
            subState.on("value_changed", _updateEffectiveTransitionTime.bind(null, self, subState));
        }
    }
}
exports.UATwoStateVariable = UATwoStateVariable;
function _addTwoStateVariable(namespace, options) {
    const addressSpace = namespace.addressSpace;
    const twoStateVariableType = addressSpace.findVariableType("TwoStateVariableType");
    if (!twoStateVariableType) {
        throw new Error("cannot find TwoStateVariableType");
    }
    options.optionals = options.optionals || [];
    if (options.trueState) {
        options.optionals.push("TrueState");
    }
    if (options.falseState) {
        options.optionals.push("FalseState");
    }
    // we want event based changes...
    options.minimumSamplingInterval = 0;
    const node = twoStateVariableType.instantiate({
        browseName: options.browseName,
        nodeId: options.nodeId,
        description: options.description,
        componentOf: options.componentOf,
        organizedBy: options.organizedBy,
        modellingRule: options.modellingRule,
        dataType: node_opcua_nodeid_1.resolveNodeId(node_opcua_variant_2.DataType.LocalizedText),
        minimumSamplingInterval: options.minimumSamplingInterval,
        optionals: options.optionals
    });
    const _node = _install_TwoStateVariable_machinery(node, options);
    return _node;
}
exports._addTwoStateVariable = _addTwoStateVariable;
