// node-opcua-address-space/dist/src/data_access/ua_multistate_value_discrete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._addMultiStateValueDiscrete = exports.promoteToMultiStateValueDiscrete = exports.UAMultiStateValueDiscrete = void 0;
/**
 * @module node-opcua-address-space.DataAccess
 */
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_variant_1 = require("node-opcua-variant");
const node_opcua_basic_types_1 = require("node-opcua-basic-types");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_status_code_1 = require("node-opcua-status-code");
const ua_variable_1 = require("../ua_variable");
const node_opcua_constants_1 = require("node-opcua-constants");
const register_node_promoter_1 = require("../../source/loader/register_node_promoter");
const coerce_enum_value_1 = require("../../source/helpers/coerce_enum_value");
const ua_data_item_1 = require("./ua_data_item");
function install_synchronization(variable) {
    variable.on("value_changed", (value) => {
        const valueAsTextNode = variable.valueAsText || variable.getComponentByName("ValueAsText");
        if (!valueAsTextNode) {
            return;
        }
        const valueAsText1 = variable._findValueAsText(value.value.value);
        valueAsTextNode.setValueFromSource(valueAsText1);
    });
    variable.emit("value_changed", variable.readValue());
}
class UAMultiStateValueDiscrete extends ua_variable_1.UAVariable {
    setValue(value) {
        if (typeof value === "string") {
            const enumValues = this.enumValues.readValue().value.value;
            const selected = enumValues.filter((a) => a.displayName.text === value)[0];
            if (selected) {
                this._setValue(selected.value);
            }
            else {
                throw new Error("cannot find enum string " + value + " in " + enumValues.toString());
            }
        }
        else {
            this._setValue(node_opcua_basic_types_1.coerceUInt64(value));
        }
    }
    getValueAsString() {
        return this.valueAsText.readValue().value.value.text || "";
    }
    getValueAsNumber() {
        return this.readValue().value.value;
    }
    checkVariantCompatibility(value) {
        if (this.enumValues) {
            if (!this._isValueInRange(node_opcua_basic_types_1.coerceInt32(value.value))) {
                return node_opcua_status_code_1.StatusCodes.BadOutOfRange;
            }
        }
        return node_opcua_status_code_1.StatusCodes.Good;
    }
    clone(options1, optionalFilter, extraInfo) {
        const variable1 = ua_variable_1.UAVariable.prototype.clone.call(this, options1, optionalFilter, extraInfo);
        return promoteToMultiStateValueDiscrete(variable1);
    }
    /**
     * @private
     */
    _isValueInRange(value) {
        // MultiStateValueDiscreteType
        const enumValues = this.enumValues.readValue().value.value;
        const e = enumValues.findIndex((x) => node_opcua_basic_types_1.coerceInt32(x.value) === value);
        return !(e === -1);
    }
    /**
     *
     * @private
     */
    _enumValueIndex() {
        // construct an index to quickly find a EnumValue from a value
        const enumValues = this.enumValues.readValue().value.value;
        const enumValueIndex = {};
        if (!enumValues || !enumValues.forEach) {
            return enumValueIndex;
        }
        enumValues.forEach((e) => {
            enumValueIndex[e.value[1]] = e;
        });
        return enumValueIndex;
    }
    /**
     *
     * @private
     */
    _setValue(value) {
        // check that value is in bound
        if (!this._isValueInRange(node_opcua_basic_types_1.coerceInt32(value))) {
            throw new Error("UAMultiStateValueDiscrete#_setValue out of range " + value);
        }
        const dataType = this._getDataType();
        if (dataType === node_opcua_variant_1.DataType.Int64 || dataType === node_opcua_variant_1.DataType.UInt64) {
            this.setValueFromSource({ dataType, value });
        }
        else {
            const valueN = value[1];
            this.setValueFromSource({ dataType, value: valueN });
        }
    }
    /**
     *
     * @private
     */
    _findValueAsText(value) {
        const enumValueIndex = this._enumValueIndex();
        if (value === undefined) {
            throw new Error("Unexpected undefined value");
        }
        if (value instanceof Array) {
            value = value[1];
        }
        node_opcua_assert_1.assert(!(value instanceof node_opcua_variant_1.Variant));
        let valueAsText1 = "Invalid";
        if (enumValueIndex[value]) {
            valueAsText1 = enumValueIndex[value].displayName;
        }
        const result = new node_opcua_variant_1.Variant({
            dataType: node_opcua_variant_1.DataType.LocalizedText,
            value: node_opcua_data_model_1.coerceLocalizedText(valueAsText1)
        });
        return result;
    }
    _getDataType() {
        if (this.dataType.value === 26 /* Number */) {
            return node_opcua_variant_1.DataType.UInt32;
        }
        const dataTypeStr = node_opcua_variant_1.DataType[this.dataType.value];
        return node_opcua_variant_1.DataType[dataTypeStr];
    }
    /**
     *
     * @private
     */
    _post_initialize() {
        // MultiStateValueDiscrete Variables can have any numeric Data Type;
        // this includes signed and unsigned integers from 8 to 64 Bit length.
        // istanbul ignore next
        if (typeof this.dataType.value !== "number" ||
            [
                node_opcua_variant_1.DataType.UInt64,
                node_opcua_variant_1.DataType.Int64,
                node_opcua_variant_1.DataType.UInt32,
                node_opcua_variant_1.DataType.Int32,
                node_opcua_variant_1.DataType.UInt16,
                node_opcua_variant_1.DataType.Int16,
                node_opcua_variant_1.DataType.Byte,
                node_opcua_variant_1.DataType.Byte,
                node_opcua_variant_1.DataType.SByte,
                26 /*Number*/
            ].indexOf(this.dataType.value) <= 0) {
            throw new Error("Invalid DataType in UAMultiStateValueDiscrete =>" + this.dataType.toString());
        }
        // find the enum value type
        install_synchronization(this);
    }
}
exports.UAMultiStateValueDiscrete = UAMultiStateValueDiscrete;
function promoteToMultiStateValueDiscrete(node) {
    if (node instanceof UAMultiStateValueDiscrete) {
        return node; // already promoted
    }
    Object.setPrototypeOf(node, UAMultiStateValueDiscrete.prototype);
    node_opcua_assert_1.assert(node instanceof UAMultiStateValueDiscrete, "should now  be a State Machine");
    node._post_initialize();
    return node;
}
exports.promoteToMultiStateValueDiscrete = promoteToMultiStateValueDiscrete;
register_node_promoter_1.registerNodePromoter(node_opcua_constants_1.VariableTypeIds.MultiStateValueDiscreteType, promoteToMultiStateValueDiscrete);
function _addMultiStateValueDiscrete(namespace, options) {
    node_opcua_assert_1.assert(options.hasOwnProperty("enumValues"));
    node_opcua_assert_1.assert(!options.hasOwnProperty("ValuePrecision"));
    const addressSpace = namespace.addressSpace;
    const multiStateValueDiscreteType = addressSpace.findVariableType("MultiStateValueDiscreteType");
    if (!multiStateValueDiscreteType) {
        throw new Error("expecting MultiStateValueDiscreteType to be defined , check nodeset xml file");
    }
    // todo : if options.typeDefinition is specified, check that type is SubTypeOf MultiStateDiscreteType
    // EnumValueType
    //   value: Int64, displayName: LocalizedText, Description: LocalizedText
    const enumValues = coerce_enum_value_1.coerceEnumValues(options.enumValues);
    if (options.value === undefined && enumValues[0]) {
        options.value = enumValues[0].value; // Int64
    }
    let value;
    if (typeof options.value === "number" || node_opcua_basic_types_1.isValidUInt64(options.value)) {
        if (node_opcua_basic_types_1.isValidUInt64(options.value)) {
            value = new node_opcua_variant_1.Variant({
                dataType: node_opcua_variant_1.DataType.UInt32,
                value: options.value[1] // Low word
            });
        }
        else {
            value = new node_opcua_variant_1.Variant({
                dataType: node_opcua_variant_1.DataType.UInt32,
                value: options.value
            });
        }
    }
    else {
        value = options.value;
    }
    const cloned_options = Object.assign(Object.assign({}, options), { dataType: node_opcua_variant_1.DataType.UInt32, typeDefinition: multiStateValueDiscreteType.nodeId, 
        // valueRank:
        // note : OPCUA Spec 1.03 specifies -1:Scalar (part 8 page 8) but nodeset file specifies -2:Any
        value, valueRank: -1 // -1 : Scalar
     });
    const variable = namespace.addVariable(cloned_options);
    ua_data_item_1.add_dataItem_stuff(variable, options);
    namespace.addVariable({
        accessLevel: "CurrentRead",
        browseName: { name: "EnumValues", namespaceIndex: 0 },
        dataType: "EnumValueType",
        minimumSamplingInterval: 0,
        modellingRule: options.modellingRule ? "Mandatory" : undefined,
        propertyOf: variable,
        typeDefinition: "PropertyType",
        userAccessLevel: "CurrentRead",
        value: new node_opcua_variant_1.Variant({
            arrayType: node_opcua_variant_1.VariantArrayType.Array,
            dataType: node_opcua_variant_1.DataType.ExtensionObject,
            value: enumValues
        })
    });
    namespace.addVariable({
        accessLevel: "CurrentRead",
        browseName: { name: "ValueAsText", namespaceIndex: 0 },
        dataType: node_opcua_variant_1.DataType.LocalizedText,
        minimumSamplingInterval: 0,
        modellingRule: options.modellingRule ? "Mandatory" : undefined,
        propertyOf: variable,
        typeDefinition: "PropertyType",
        userAccessLevel: "CurrentRead"
        // value: valueAsText
    });
    // install additional helpers methods
    variable.install_extra_properties();
    promoteToMultiStateValueDiscrete(variable);
    node_opcua_assert_1.assert(variable.enumValues.browseName.toString() === "EnumValues");
    node_opcua_assert_1.assert(variable.valueAsText.browseName.toString() === "ValueAsText");
    return variable;
}
exports._addMultiStateValueDiscrete = _addMultiStateValueDiscrete;
